### PR TITLE
Converted service to PR

### PR DESCRIPTION
**Pull request by vlf-stefan-drobic**
As per @fedjo 's suggestion, I've created a PR containing the curent implementation of this service up until this moment in time.
The main branch is currently empty, containing a readme file with the name of the service.

@fedjo we can continue the discussion you had in mind in this PR, then commit new changes and push to main once it is deemed satisfactory.

@arruda @nikoskal

### CODE REVIEW

File: app/init/db_init.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        db.flush()
+        db.close()
+    except Exception as e:
+        print(e)
+        print("User already exists, fine to continue.")
+
+    print("init_db done")
+
+
+if __name__ == "__main__":
```
Comment (fedjo) [inline app/init/db_init.py:37]: maybe remove this as we are not running the file as a script

**Comment by fedjo** [comment]
Very nice code! 

As I see you have used the same skeleton from PND service, please as general guidelines

- remove comments and unesseray code
- replace prints with logging
- remove passwords
- if needed move automigrations outside of the app

File: defaults.env [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# DB
+POSTGRES_USER=postgres
+POSTGRES_PASSWORD=postgres
+POSTGRES_DB=postgres
+# If the HOST var changes, it also needs to be manually changed in the alembic.ini file (This should only be changed if
+# the hostname of the db container is changed)
+POSTGRES_HOST=postgres
+# POSTGRES_DB_PORT=5432
```
Comment (fedjo) [inline defaults.env:8]: why commented out?

File: README.md
Comment (fedjo) [inline README.md:1]: Please create a section where you provide an example request and response payload for your API calls.

If they are too large and make the REAMDE difficult to read, please consider creating another APIcalls.md file with this information

File: app/api/api_v1/endpoints/data.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+from schemas import DataCreate, DataID, Message
+
+router = APIRouter()
+
+
+@router.get("/{dataset_id}")
+def get_by_id(
+        dataset_id: int,
+        current_user: User = Depends(deps.get_current_user),
```
Comment (fedjo) [inline app/api/api_v1/endpoints/data.py:19]: why do you requiere `current_user` ?

File: app/api/api_v1/endpoints/report.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+    report_db = crud.report.get(db=db, id=report_id)
+
+    if not report_db:
+        raise HTTPException(
+            status_code=400,
+            detail="Report with ID:{} does not exist.".format(report_id)
+        )
+
+    types = {"work_book": "work-book", "plant_protection": "plant-protection", "irrigation": "irrigations", "fertilisations": "fertilisations", "harvest": "harvests", "global_gap": "GlobalGAP"}
```
Comment (fedjo) [inline app/api/api_v1/endpoints/report.py:37]: is this variable used?

File: app/api/api_v1/endpoints/report.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+from fastapi import APIRouter, Depends, Response, HTTPException
+from fpdf import FPDF
+from sqlalchemy.orm import Session
+
+import crud
+import utils
+from api import deps
+from models import User
+from schemas import FarmProfile, MachineryAssetsOfFarm, PlotParcelDetail, ReportDB, ReportCreate, Message, ReportDBID
```
Comment (fedjo) [inline app/api/api_v1/endpoints/report.py:12]: Are these schemas used?

File: app/schemas/report.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    name: str
+
+
+class ReportDBID(BaseModel):
+    model_config = ConfigDict(from_attributes=True)
+
+    id: int
+
+
+class FarmProfile(BaseModel):
```
Comment (fedjo) [inline app/schemas/report.py:30]: What is the purpose of these schemas and where are they used?

File: tests/tests_.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
+import pytest
+
+from app.main import app
+from fastapi.testclient import TestClient
+
+client = TestClient(app)
+
+
+@pytest.fixture()
```
Comment (fedjo) [inline tests/tests_.py:9]: consider creating fixture in a separate `fixtures.py` file

File: testing.env
Comment (fedjo) [inline testing.env:1]: is this file needed?

File: app/utils/utils.py
Comment (fedjo) [inline app/utils/utils.py:1]: This is a very long file. Please consider adding a utils directory.
Eg

```
utils/
    pdf_utils.py
    parse_utils.py
```

or

```
utils/
   fertilization_report_utils.py
   irrigation_report_utils.py
```

File: app/api/api_v1/endpoints/report.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
+import datetime
+import json
+
+from fastapi import APIRouter, Depends, Response, HTTPException
+from fpdf import FPDF
+from sqlalchemy.orm import Session
+
+import crud
+import utils
```
Comment (fedjo) [inline app/api/api_v1/endpoints/report.py:9]: `utils` is imported twice

File: app/api/api_v1/endpoints/report.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            plot=utils.parse_plot_detail(json_file),
+            cult=utils.parse_generic_cultivation_info(json_file),
+            irri=utils.parse_irrigation(json_file),
+            fert=utils.parse_fertilization(json_file),
+            pdmd=utils.parse_plant_protection(json_file)
+        )
+
+    elif report_db.type == "plant-protection":
+
+        pdf = plant_protection(utils.parse_plant_protection(json.loads(report_db.file)))
```
Comment (fedjo) [inline app/api/api_v1/endpoints/report.py:53]: better to use it as `utils.plant_protection(...`

File: app/api/api_v1/endpoints/report.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+@router.post("/{report_type}/dataset/{dataset_id}", response_model=ReportDBID)
+def create_by_data_id(
+        dataset_id: int,
+        report_type: str,
+        current_user: User = Depends(deps.get_current_user),
+        db: Session = Depends(deps.get_db)
+) -> ReportDBID:
+    """
+    Generate a report based off of a previously uploaded/queried data file.
```
Comment (fedjo) [inline app/api/api_v1/endpoints/report.py:95]: typo

File: app/api/api_v1/endpoints/report.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        )
+
+    if report_type not in types.values():
+        raise HTTPException(
+            status_code=400,
+            detail="Report type {} isn't part of the offered types.".format(report_type)
+        )
+
+    # Create DB entry
+    report_db = crud.report.create(db=db, obj_in=ReportCreate(name=dataset_db.filename + " report.pdf", file=dataset_db.data, type=report_type))
```
Comment (fedjo) [inline app/api/api_v1/endpoints/report.py:118]: why are you saving in the `Report.file` the `Data.file`? In this way you save the same file in the DB twice, I am right? I would expect to save in the db the generated pdf file of the report.

File: app/api/api_v1/endpoints/report.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            status_code=400,
+            detail="Report with ID:{} does not exist.".format(report_id)
+        )
+
+    types = {"work_book": "work-book", "plant_protection": "plant-protection", "irrigation": "irrigations", "fertilisations": "fertilisations", "harvest": "harvests", "global_gap": "GlobalGAP"}
+
+    if report_db.type == "work-book":
+        json_file = json.loads(report_db.file)
+
+        pdf = utils.work_book(
```
Comment (fedjo) [inline app/api/api_v1/endpoints/report.py:42]: I see that you are creating the pdf reports on the fly every time they are requested. This is ok for the project now, but in the future we would like to save in the `Report` model the pdf file and when requested to just serve the file.

In that way we create the pdf on upload (offline) and the API will be a lot faster!

File: app/init/db_init.py (hunk 1 above)
Comment (vlf-stefan-drobic) [reply app/init/db_init.py:37]: Removed the init_db completely, new users simply register on the service

File: defaults.env (hunk 2 above)
Comment (vlf-stefan-drobic) [reply defaults.env:8]: Refactored to be like the pdm service

File: README.md
Comment (vlf-stefan-drobic) [reply README.md:1]: Added examples for the API calls

File: app/api/api_v1/endpoints/data.py (hunk 3 above)
Comment (vlf-stefan-drobic) [reply app/api/api_v1/endpoints/data.py:19]: Authentication, if the user is not logged in, then they won't have access to the api

File: app/api/api_v1/endpoints/report.py (hunk 4 above)
Comment (vlf-stefan-drobic) [reply app/api/api_v1/endpoints/report.py:37]: Nope, leftover when I was designing the APIs, removed

File: app/api/api_v1/endpoints/report.py (hunk 5 above)
Comment (vlf-stefan-drobic) [reply app/api/api_v1/endpoints/report.py:12]: Some are, removed the ones that weren't being used

File: testing.env
Comment (vlf-stefan-drobic) [reply testing.env:1]: I thought it was a good idea to create separate a separate .env file for testing purposes, but it really doesn't simplify much at the moment, so I've removed it

File: app/utils/utils.py [hunk 13]
```diff
@@ -0,0 +1,842 @@
...
+            row.cell(key)
+        for i in range(1):
+            row = table.row()
+            for j in range(4):
+                row.cell(None)
+
+    # 3
+
+    first = 1
+    first_page = True
```
[the remark this answers is not in the document]
Comment (vlf-stefan-drobic) [reply app/utils/utils.py:599]: They are used when adding plot information, because the plots are represented using images, and the images can be huge at points, I keep a count of what page (image) I'm on, so as to not cause either an image overlap, or have more than one image on a page.

File: app/utils/utils.py
Comment (vlf-stefan-drobic) [reply app/utils/utils.py:1]: I'd prefer to open a separate issue to refactor this, because I don't want this PR to become crowded

File: app/api/api_v1/endpoints/report.py (hunk 8 above)
Comment (vlf-stefan-drobic) [reply app/api/api_v1/endpoints/report.py:9]: Removed second import

File: app/api/api_v1/endpoints/report.py (hunk 9 above)
Comment (vlf-stefan-drobic) [reply app/api/api_v1/endpoints/report.py:53]: Agreed, updated the usage for the utils functions

File: app/api/api_v1/endpoints/report.py (hunk 10 above)
Comment (vlf-stefan-drobic) [reply app/api/api_v1/endpoints/report.py:95]: I'm not seeing it and my IDE isn't reporting it
I did update the text to better describe the api

File: app/api/api_v1/endpoints/report.py (hunk 12 above)
Comment (vlf-stefan-drobic) [reply app/api/api_v1/endpoints/report.py:42]: Agreed, as the reports are relatively small and take a small amount of work to be created, it's fast enough. But I'll make a feature issue to change to a store the report in the database and fetch when needed method.

File: app/api/api_v1/endpoints/report.py (hunk 11 above)
Comment (vlf-stefan-drobic) [reply app/api/api_v1/endpoints/report.py:118]: The report can be saved to the database instead of generating on the fly in the API, I choose this way because it's faster, for now at least.
If a huge report is requested, or one that contains a lot of data, then this method will be slower instead of caching the report and fetching it from the db when needed.
I'd make a separate feature issue of this and implement the changes in the coming weeks

**Comment by vlf-stefan-drobic** [comment]
I went over all of the previous comments, and marked them as resolved, @fedjo do take another look

File: app/api/api_v1/endpoints/report.py (hunk 11 above)
Comment (fedjo) [reply app/api/api_v1/endpoints/report.py:111]: Ok. Let's create a separate issue for saving the created pdf report file to the DB.

In any case, saving `dataset.data` to `Report.file` seems a big data duplication on the DB.

File: app/core/config.py [hunk 14]
```diff
@@ -49,9 +45,7 @@ def assemble_db_connection(cls, v: Optional[str], values) -> Any:
         .has().no().spaces() \
 
     ACCESS_TOKEN_EXPIRATION_TIME: int
-    KEY: str = "c2bab29d257f0ffc52d9ac677d4ff6d1d9d5e92e3d3939d3f4cwc"
-
-    # model_config = SettingsConfigDict(case_sensitive=True, env_file="defaults.env")
+    JWT_KEY: str = "c2bab29d257f0ffc52d9ac677d4ff6d1d9d5e92e3d3939d3f4cwc"
```
Comment (fedjo) [inline app/core/config.py:48]: Hmmm maybe this is a good idea to add it in the `.env` file, since when using the gatekeeper the JWT_KEY should be the same one used in the gatekeeper, so that the service can decode the JWTtoken sent by the gatekeeper.

**Review by fedjo — DISMISSED** [review]

**Comment by fedjo** [comment]
For me, if support for Unicode names is done, PR is approved!

**Review by fedjo — APPROVED** [review]

**Comment by vlf-stefan-drobic** [comment]
Alright, support for unicode characters has been added with the introduction of a unicode friendly font.
I added the freeserif font from here: https://www.gnu.org/software/freefont/
@fedjo do you think I need to add additional information about this in the readme, the fonts are "free" as in free beer and are covered under the GNU GPL licence like other pip libraries

**Comment by fedjo** [comment]
> Alright, support for unicode characters has been added with the introduction of a unicode friendly font. I added the freeserif font from here: https://www.gnu.org/software/freefont/ @fedjo do you think I need to add additional information about this in the readme, the fonts are "free" as in free beer and are covered under the GNU GPL licence like other pip libraries

Excellent! 

I added just a small comment.

Regarding fonts, I do not believe you should mention anything. Let's have also the opinion of @arruda

File: app/utils/FreeSerif.ttf
Comment (fedjo) [inline app/utils/FreeSerif.ttf:1]: Better include them in a `fonts/` folder. Either in `utils/` directory or in a `static/` directory

**Review by fedjo — APPROVED** [review]
